### PR TITLE
[type-checker] More correct way to determine if a given generic parameter of a function cannot be inferred because it is not used

### PR DESCRIPTION
**Pull request by swiftix**
The overall idea is like this:
- first, collect the generic params that are explicitly used by types of arguments or return type
- then perform a fix-point analysis, where at each step, a set of generic signature’s requirements is checked.
   - If there is a requirement where one of the involved generic params is already proven to be used, then all other generic params from the same requirement are also considered to be implicitly used in the function signature, because they are used to put a requirement on a generic param that is already known to be used in the function signature.

The fix-point iteration is required to handle chains of dependencies between requirements. See the updated test case for an example.

Fixes rdar://29585211

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by slavapestov** [comment]
This looks fine, but as I understand it not diagnosing this case just leads to a type checked function that can't be called, but is otherwise harmless. Were you seeing crashes or something?

**Comment by swiftix** [comment]
@slavapestov No, there are no crashes due to this. The function just cannot be called. Plus, @moiseev  was complaining about this, because he has written such functions a couple of times and could not understand why they were not called and a different function with the same name was picked.

**Comment by swiftix** [comment]
Fun. It has found a bunch of places where the generic param cannot be inferred. We haven't noticed it ourselves ;-)

**Comment by swiftix** [comment]
@DougGregor Doug, WDYT? In particular, could you open the Details to see the failures, which are the newly identified places (about a dozen such places), where we had generic parameters that cannot be ever inferred at call-site. Can you confirm that that they are identified correctly and not a bug in my implementation? May be I misunderstood something? E.g. I assumed that knowing T.AssocType is not enough to infer T itself at the call-site.

**Comment by slavapestov** [comment]
@swiftix You are correct that knowing T.AssocType is not enough to know T without further constraints.

File: lib/Sema/TypeCheckGeneric.cpp [hunk 1]
```diff
@@ -616,32 +616,124 @@ static void checkReferencedGenericParams(GenericContext *dc,
...
-
-      case RequirementKind::Conformance:
-      case RequirementKind::Layout:
-        req.getFirstType().visit(visitorFn);
-        break;
+    case RequirementKind::SameType:
+      second = req.getSecondType();
+      LLVM_FALLTHROUGH;
+
+    case RequirementKind::Superclass:
```
Comment (DougGregor) [inline lib/Sema/TypeCheckGeneric.cpp]: Shouldn't we also capture the second type of superclass constraints, if only for future-correctness?

Comment (swiftix) [inline lib/Sema/TypeCheckGeneric.cpp]: Yeah. I can easily do it.

Comment (swiftix) [inline lib/Sema/TypeCheckGeneric.cpp]: Any further comments regarding this PR? Are those new cases that it caught (see https://ci.swift.org/job/swift-PR-Linux/6568/) correctly identified as bugs?

File: lib/Sema/TypeCheckGeneric.cpp [hunk 2]
```diff
@@ -616,32 +616,124 @@ static void checkReferencedGenericParams(GenericContext *dc,
...
+                    [&referencedGenericParams,
+                     &FoundNewReferencedGenericParam](CanType t) {
+                      // Add only generic type parameters, but ignore any
+                      // dependent member types, because requirement
+                      // on a dependent member type does not provide enough
+                      // information to infer the base generic type
+                      // parameter.
+                      if (!t->is<GenericTypeParamType>())
+                        return;
+                      FoundNewReferencedGenericParam |=
```
Comment (DougGregor) [inline lib/Sema/TypeCheckGeneric.cpp]: I don't think you want the short-circuiting here. I suggest:

    if (referencedGenericParams.insert(t).second) 
      FoundNewReferencedGenericParam = true;

File: lib/Sema/TypeCheckGeneric.cpp [hunk 3]
```diff
@@ -616,32 +616,124 @@ static void checkReferencedGenericParams(GenericContext *dc,
...
+  // not-yet-referenced-T1 == not-yet-referenced-T2.DepType2,
+  // not-yet-referenced-T2 == not-yet-referenced-T3.DepType3,
+  // not-yet-referenced-T3 == referenced-T4.DepType4.
+  // When we process the first of these requirements, we don't know yet that T2
+  // will be referenced, because T3 will be referenced,
+  // because T3 == T4.DepType4.
+  while (true) {
+    bool FoundNewReferencedGenericParam = false;
+    for (auto req : requirements) {
+      FoundNewReferencedGenericParam |= reqTypesVisitor(req);
```
Comment (DougGregor) [inline lib/Sema/TypeCheckGeneric.cpp]: Same short-circuiting issue here; we want to record everything we see, I think.

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@swift-ci please smoke test

**Comment by swiftix** [comment]
@DougGregor I updated the tests and addressed all of your comments. Looks OK now?

**Comment by swiftix** [comment]
@swift-ci please test

**Comment by swiftix** [comment]
@swift-ci please test

**Pull request merged**